File: src/environment/dependencyInventory.ts
export interface DependencyInventory {
    schemaVersion: 'dependency-inventory-v1';
    filesScanned: number; excludedDirectories: number; complete: boolean; dynamicImports: number;
    imports: { module: string; kind: 'external' | 'stdlib' | 'local';
        availability: 'available' | 'missing' | 'unknown' | 'not-checked';
        references: { file: string; line: number; context: 'required' | 'optional' | 'typing' | 'conditional' }[] }[];
    issues: { file: string; reason: string }[];
    missing: string[]; optionalMissing: string[];
}

export function isDependencyInventory(value: unknown): value is DependencyInventory {
    if (!value || typeof value !== 'object') { return false; }
    const scan = value as DependencyInventory;
    const count = (value: unknown) => Number.isSafeInteger(value) && (value as number) >= 0;
    const text = (value: unknown) => typeof value === 'string' && value.length <= 4096 && !/[\r\n\0]/.test(value);
    const names = (value: unknown) => Array.isArray(value) && value.length <= 10000
        && value.every(name => typeof name === 'string' && /^[\p{L}_][\p{L}\p{N}_]*$/u.test(name));
    return scan.schemaVersion === 'dependency-inventory-v1' && typeof scan.complete === 'boolean'
        && count(scan.filesScanned) && count(scan.excludedDirectories) && count(scan.dynamicImports)
        && names(scan.missing) && names(scan.optionalMissing)
        && Array.isArray(scan.issues) && scan.issues.length <= 10100
        && scan.issues.every(issue => issue && text(issue.file) && text(issue.reason))
        && Array.isArray(scan.imports) && scan.imports.length <= 10000
        && scan.imports.every(item => item && text(item.module) && ['external', 'stdlib', 'local'].includes(item.kind)
            && ['available', 'missing', 'unknown', 'not-checked'].includes(item.availability)
            && Array.isArray(item.references) && item.references.length <= 10000 && item.references.length > 0
            && item.references.every(ref => ref && text(ref.file) && count(ref.line) && ref.line > 0
                && ['required', 'optional', 'typing', 'conditional'].includes(ref.context)));
}

export function inventorySummary(scan: DependencyInventory): string {
    return `已掃描 ${scan.filesScanned} 個 Python 檔案、${scan.imports.length} 種 import；缺少 ${scan.missing.length} 個必要套件，`
        + `${scan.optionalMissing.length} 個條件／可選套件；掃描${scan.complete ? '完成' : '不完整'}。`;
}

/** Only module identifiers, project-relative references and stable diagnostic codes. */
export function inventoryReport(scan: DependencyInventory, initialMissing: string[] = [], outcome = '僅完成靜態相依盤點。'): string {
    const cell = (value: string) => value.replace(/[&<>|`\r\n]/g, char => `&#${char.charCodeAt(0)};`);
    const kinds = { external: '外部套件', stdlib: '標準庫', local: '專案模組' };
    const statuses = { available: '可找到頂層套件', missing: '缺少', unknown: '無法判定', 'not-checked': '未驗證載入' };
    const contexts = { required: '必要', optional: '可選', typing: '型別檢查', conditional: '條件分支' };
    return ['# Python 相依掃描', '', cell(outcome), '', inventorySummary(scan), '',
        '此報告只解析所選範圍的 import 並檢查頂層套件是否存在，不執行專案或外部套件。套件版本、子模組、載入副作用與動態相依仍須由正式測試預檢驗證。', '',
        `首次缺少：${initialMissing.map(cell).join('、') || '無'}。`,
        `目前必要套件缺少：${scan.missing.map(cell).join('、') || '無'}。`,
        `條件／可選／型別套件缺少：${scan.optionalMissing.map(cell).join('、') || '無'}；列入報告，不自動補裝。`,
        `略過 ${scan.excludedDirectories} 個環境、建置、輸出或連結目錄；發現 ${scan.dynamicImports} 處常見動態 import 呼叫。計算得出的或別名動態 import 無法完整靜態辨識。`, '',
        ...scan.issues.map(issue => `- ${cell(issue.file)}：${cell(issue.reason)}`), '',
        '| Import | 分類 | 檢查結果 | 使用位置（所選範圍相對路徑） |', '| --- | --- | --- | --- |',
        ...scan.imports.map(item => `| ${cell(item.module)} | ${kinds[item.kind]} | ${statuses[item.availability]} | `
            + item.references.map(ref => `${cell(ref.file)}:${ref.line}（${contexts[ref.context]}）`).join('<br>') + ' |'), '',
        '缺少必要套件時優先使用所選目錄至專案根目錄間最近的 requirements；沒有清單時只採 llmUnitTest.packageMappings 的明確對應。語法錯誤、讀取失敗或容量超限均不能算掃描成功。', ''].join('\n');
}

File: python_scripts/dependency_inventory.py
"""Bounded AST import inventory. Never import application or dependency modules."""
import ast
import importlib.machinery
import json
import os
from pathlib import Path
import sys
import tokenize


EXCLUDED = {'.git', '.hg', '.svn', '.venv', 'venv', 'env', 'node_modules',
            '__pycache__', '.pytest_cache', '.mypy_cache', '.tox', '.nox',
            'site-packages', 'dist', 'build', '.eggs'}
MAX_FILES = 5000
MAX_BYTES = 32 * 1024 * 1024
MAX_FILE_BYTES = 2 * 1024 * 1024
MAX_REFERENCES = 10000


def generated_directory(directory):
    """Recognize our artifacts by their manifest, not arbitrary user folder names."""
    manifest = directory / 'run_manifest.json'
    try:
        if manifest.is_file() and manifest.stat().st_size < 65536:
            value = json.loads(manifest.read_text(encoding='utf-8-sig'))
            return (isinstance(value, dict) and value.get('schemaVersion') == 2 and 'runId' in value
                    and 'sourceHash' in value and str(value.get('promptVersion', '')).startswith('role-contracts-'))
    except (OSError, ValueError):
        pass
    return False


class ImportVisitor(ast.NodeVisitor):
    def __init__(self):
        self.imports = []
        self.context = 'required'
        self.dynamic_calls = 0

    def visit_Import(self, node):
        self.imports.extend((alias.name, node.lineno, self.context) for alias in node.names)

    def visit_ImportFrom(self, node):
        self.imports.append(('.' * node.level + (node.module or ''), node.lineno, self.context))

    def visit_Call(self, node):
        # Aliased/computed dynamic imports cannot be exhaustively discovered statically.
        if ((isinstance(node.func, ast.Name) and node.func.id == '__import__') or
                (isinstance(node.func, ast.Attribute) and node.func.attr == 'import_module')):
            self.dynamic_calls += 1
        self.generic_visit(node)

    def visit_If(self, node):
        previous = self.context
        typing = ((isinstance(node.test, ast.Name) and node.test.id == 'TYPE_CHECKING') or
                  (isinstance(node.test, ast.Attribute) and node.test.attr == 'TYPE_CHECKING'))
        self.visit(node.test)
        self.context = 'typing' if typing else ('conditional' if previous == 'required' else previous)
        for child in node.body:
            self.visit(child)
        self.context = previous if typing else self.context
        for child in node.orelse:
            self.visit(child)
        self.context = previous

    def visit_Try(self, node):
        previous = self.context
        catches_import = any(handler.type and any(isinstance(part, ast.Name) and
                             part.id in ('ImportError', 'ModuleNotFoundError')
                             for part in ast.walk(handler.type)) for handler in node.handlers)
        if catches_import and previous == 'required':
            self.context = 'optional'
        for child in node.body:
            self.visit(child)
        self.context = 'conditional' if previous == 'required' else previous
        for handler in node.handlers:
            self.visit(handler)
        self.context = previous
        for child in [*node.orelse, *node.finalbody]:
            self.visit(child)


def inventory(scan_root, project_root=None, excluded_paths=()):
    root = Path(scan_root).resolve()
    project = Path(project_root or root).resolve()
    if not root.is_relative_to(project):
        project = root
    excluded_paths = [Path(value).resolve() for value in excluded_paths if value]
    result = {'schemaVersion': 'dependency-inventory-v1', 'filesScanned': 0,
              'excludedDirectories': 0, 'complete': True, 'dynamicImports': 0,
              'imports': [], 'issues': [], 'missing': [], 'optionalMissing': []}
    records = {}
    total_bytes = 0
    references = 0
    files_seen = 0

    def issue(file, reason):
        result['complete'] = False
        if len(result['issues']) < 100:
            result['issues'].append({'file': file, 'reason': reason})

    if not root.is_dir() or (root / 'pyvenv.cfg').is_file():
        issue('.', 'invalid-source-directory')
        return result
    def walk_error(error):
        # Do not expose arbitrary OS errors or source text.
        issue('.', 'directory-unreadable')

    for directory, dirs, files in os.walk(root, followlinks=False, onerror=walk_error):
        folder = Path(directory)
        if generated_directory(folder) or (folder / 'pyvenv.cfg').is_file():
            dirs[:] = []
            result['excludedDirectories'] += 1
            continue
        kept = []
        for name in sorted(dirs):
            child = folder / name
            if (name.lower() in EXCLUDED or child.is_symlink() or
                    (hasattr(child, 'is_junction') and child.is_junction()) or
                    not child.resolve().is_relative_to(root) or
                    any(child.resolve() == excluded for excluded in excluded_paths)):
                result['excludedDirectories'] += 1
            else:
                kept.append(name)
        dirs[:] = kept
        for name in sorted(files):
            if not name.endswith('.py'):
                continue
            source = folder / name
            files_seen += 1
            relative = source.relative_to(root).as_posix()
            if source.is_symlink() or not source.resolve().is_relative_to(root):
                issue(relative, 'linked-source-skipped')
                continue
            if files_seen > MAX_FILES or total_bytes >= MAX_BYTES or references >= MAX_REFERENCES:
                issue(relative, 'scan-limit')
                return finish(result, records)
            try:
                size = source.stat().st_size
                total_bytes += size
                if size > MAX_FILE_BYTES or total_bytes > MAX_BYTES:
                    issue(relative, 'source-size-limit')
                    continue
                with tokenize.open(source) as stream:
                    tree = ast.parse(stream.read(), filename=relative)
            except (SyntaxError, UnicodeError, LookupError):
                issue(relative, 'source-parse-error')
                continue
            except (OSError, ValueError, RecursionError):
                issue(relative, 'source-unreadable')
                continue
            result['filesScanned'] += 1
            visitor = ImportVisitor()
            try:
                visitor.visit(tree)
            except RecursionError:
                issue(relative, 'source-depth-limit')
                continue
            result['dynamicImports'] += visitor.dynamic_calls
            roots = [source.parent, project, project / 'src']
            plain_ancestors = []
            ancestor = source.parent
            while ancestor != project and ancestor.is_relative_to(project):
                plain_ancestors.append(ancestor)
                if (ancestor / '__init__.py').is_file():
                    roots.append(ancestor.parent)
                ancestor = ancestor.parent
            for module, line, context in visitor.imports:
                references += 1
                if references > MAX_REFERENCES:
                    issue(relative, 'scan-limit')
                    return finish(result, records)
                top = module.split('.')[0]
                local = module.startswith('.') or any((base / (top + '.py')).is_file() or
                                                       (base / top).is_dir() for base in roots)
                kind = 'local' if local else 'stdlib' if top in getattr(sys, 'stdlib_module_names', sys.builtin_module_names) else 'external'
                if kind == 'external' and any((base / (top + '.py')).is_file() or (base / top).is_dir()
                                             for base in plain_ancestors):
                    kind = 'unresolved-local'
                key = (module, kind)
                record = records.setdefault(key, {'module': module, 'kind': kind, 'references': []})
                record['references'].append({'file': relative, 'line': line, 'context': context})
    if not result['filesScanned']:
        issue('.', 'no-python-files')
    return finish(result, records)


def finish(result, records):
    availability = {}
    for key in sorted(records):
        record = records[key]
        top = record['module'].split('.')[0]
        if record['kind'] == 'external':
            if top not in availability:
                try:
                    # Top-level spec lookup never executes a package __init__ or a submodule.
                    availability[top] = 'available' if importlib.machinery.PathFinder.find_spec(top) else 'missing'
                except (ImportError, AttributeError, ValueError, OSError):
                    availability[top] = 'unknown'
            record['availability'] = availability[top]
            if record['availability'] == 'missing':
                required = any(ref['context'] == 'required' for ref in record['references'])
                result['missing' if required else 'optionalMissing'].append(top)
            if record['availability'] == 'unknown':
                result['complete'] = False
                result['issues'].append({'file': record['references'][0]['file'], 'reason': 'package-lookup-failed'})
        else:
            record['availability'] = 'not-checked'
            if record['kind'] == 'unresolved-local':
                result['complete'] = False
                result['issues'].append({'file': record['references'][0]['file'], 'reason': 'local-import-root-unresolved'})
        result['imports'].append(record)
    result['missing'] = sorted(set(result['missing']))
    result['optionalMissing'] = sorted(set(result['optionalMissing']) - set(result['missing']))
    return result
